File: ruby/spec/03_stacks_and_queues/06_animal_shelter.rb
class Animal
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def to_s
    "#{self.class.downcase}-#{name}"
  end
end

class Dog < Animal
end

class Cat < Animal
end

class AnimalContainer
  attr_reader :animal, :number

  def initialize(animal, number)
    @animal = animal
    @number = number
  end
end

class AnimalShelter
  def initialize
    @dogs = LinkedList.new
    @cats = LinkedList.new
    @number = 1
  end

  def enqueue(animal)
    container = AnimalContainer.new(animal, @number)
    @number += 1

    if animal.is_a?(Dog)
      @dogs.push(container)
    else
      @cats.push(container)
    end
  end

  def dequeue_cat
    if !@cats.empty?
      @cats.pop.animal
    end
  end

  def dequeue_dog
    if !@dogs.empty?
      @dogs.pop.animal
    end
  end

  def dequeue_any
    if !@dogs.empty? && !@cats.empty?
      if @dogs.head_value.number < @cats.head_value.number
        @dogs.pop.animal
      else
        @cats.pop.animal
      end
    elsif !@dogs.empty?
      @dogs.pop.animal
    elsif !@cats.empty?
      @cats.pop.animal
    end
  end
end

RSpec.describe AnimalShelter do
  let(:shelter) { AnimalShelter.new }

  let(:thomas) { Cat.new('Thomas') }
  let(:murka) { Cat.new('Murka') }
  let(:matroskin) { Cat.new('Matroskin') }

  let(:sharik) { Dog.new('Sharik') }
  let(:polkan) { Dog.new('Polkan') }
  let(:azor) { Dog.new('Azor') }

  describe 'enqueue - dequeue' do
    it do
      shelter.enqueue(thomas)
      shelter.enqueue(murka)
      shelter.enqueue(matroskin)

      shelter.enqueue(sharik)
      shelter.enqueue(polkan)
      shelter.enqueue(azor)

      expect(shelter.dequeue_any).to eq thomas
      expect(shelter.dequeue_dog).to eq sharik
      expect(shelter.dequeue_cat).to eq murka
      expect(shelter.dequeue_dog).to eq polkan

      expect(shelter.dequeue_any).to eq matroskin
      expect(shelter.dequeue_any).to eq azor

      expect(shelter.dequeue_any).to eq nil
    end
  end
end

File: go/03_stacks_and_queues/06_animal_shelter.go
package stacks_and_queues

const Cat = "cat"
const Dog = "dog"

type animal struct {
	kind string
	name string
}

func (a *animal) ToString() string {
	return a.kind + "-" + a.name
}

type animalContainer struct {
	animal *animal
	number int
}

// Linked List
type asNode struct {
	value *animalContainer
	next  *asNode
}

type asLinkedList struct {
	head *asNode
}

func (l *asLinkedList) Push(value *animalContainer) *asNode {
	return l.AddNode(&asNode{value: value})
}

func (l *asLinkedList) AddNode(node *asNode) *asNode {
	if l.head != nil {
		tail := l.head
		for tail.next != nil {
			tail = tail.next
		}

		tail.next = node
		return l.head
	} else {
		l.head = node
		return node
	}
}

func (l *asLinkedList) Pop() *animalContainer {
	if l.head != nil {
		value := l.head.value
		l.head = l.head.next
		return value
	} else {
		return nil
	}
}

func (l *asLinkedList) Empty() bool {
	return l.head == nil
}

func (l *asLinkedList) headValue() *animalContainer {
	if l.head != nil {
		return l.head.value
	} else {
		return nil
	}
}

type animalShelter struct {
	dogs   *asLinkedList
	cats   *asLinkedList
	number int
}

func NewAsLinkedList() *asLinkedList {
	return &asLinkedList{}
}

func NewAnimalShelter() *animalShelter {
	return &animalShelter{
		dogs:   NewAsLinkedList(),
		cats:   NewAsLinkedList(),
		number: 1,
	}
}

func (s *animalShelter) Enqueue(a *animal) {
	container := &animalContainer{animal: a, number: s.number}
	s.number += 1

	if a.kind == Dog {
		s.dogs.Push(container)
	} else {
		s.cats.Push(container)
	}
}

func (s *animalShelter) DequeueCat() *animal {
	if s.cats.Empty() == false {
		c := s.cats.Pop()
		return c.animal
	} else {
		return nil
	}
}

func (s *animalShelter) DequeueDog() *animal {
	if s.dogs.Empty() == false {
		c := s.dogs.Pop()
		return c.animal
	} else {
		return nil
	}
}

func (s *animalShelter) DequeueAny() *animal {
	if s.dogs.Empty() == false && s.cats.Empty() == false {
		if s.dogs.headValue().number < s.cats.headValue().number {
			return s.dogs.Pop().animal
		} else {
			return s.cats.Pop().animal
		}
	} else if s.dogs.Empty() == false {
		return s.dogs.Pop().animal
	} else if s.cats.Empty() == false {
		return s.cats.Pop().animal
	} else {
		return nil
	}
}
